src/java.base/share/classes/sun/security/util/PBEUtil.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright (c) 2023, Red Hat, Inc.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.util;
+
+import java.security.AlgorithmParameters;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.NoSuchAlgorithmException;
+import java.security.Provider;
+import java.security.SecureRandom;
+import java.security.spec.AlgorithmParameterSpec;
+import java.security.spec.InvalidParameterSpecException;
+import java.util.Arrays;
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.PBEParameterSpec;
+
+public final class PBEUtil {
+
+    // Used by SunJCE and SunPKCS11
+    public final static class PBES2Params {
+        private static final int DEFAULT_SALT_LENGTH = 20;
+        private static final int DEFAULT_ITERATIONS = 4096;
+
+        private int iCount;
+        private byte[] salt;
+        private IvParameterSpec ivSpec;
+
+        public IvParameterSpec getIvSpec() {
+            return ivSpec;
+        }
+
+        public AlgorithmParameters getAlgorithmParameters(int blkSize,
```

Comment:
I see most if not all of the methods in this class are lifted from other classes as the result of code refactoring. Would be nice to document its callers, e.g. where it's used.

Reply:
I added references, that include the security provider and class, to each method and inner class in PBEUtil. The instance methods in PBEUtil$PBES2Params use the reference for their containing class.

Reply:
We finally decided to include a general comment in the PBEUtil$PBES2Params class and also comments in each method. We also added comments to other PBEUtil methods.